lib/truffle/truffle/patching.rb

```diff
@@ -1,34 +1,116 @@
 module Truffle::Patching
-  TRUFFLE_PATCHES_DIRECTORY = "#{Truffle::Boot.ruby_home}/lib/patches"
-  TRUFFLE_PATCHES           = Dir.glob("#{TRUFFLE_PATCHES_DIRECTORY}/**/*.rb").
-      select { |path| File.file? path }.
-      map { |path| [path[(TRUFFLE_PATCHES_DIRECTORY.size + 1)..-4], true] }.
-      to_h
+  DIR     = "#{Truffle::Boot.ruby_home}/lib/patches"
+  # Allowed operations are: :before, :after, :instead, :ignore
+  PATCHES = { "bundler"                                 => :before,
+              "bundler/version"                         => :after,
+              "bundler/cli/exec"                        => :after,
+              "bundler/compact_index_client/updater"    => :after,
+              "bundler/current_ruby"                    => :after,
+              "bundler/dependency"                      => :after,
+              "bundler/fetcher/downloader"              => :after,
+              "bundler/fetcher/compact_index"           => :after,
+              "bundler/ruby_version"                    => :after,
+              "bundler/source/rubygems"                 => :after,
+
+              "rubygems"                                => :before,
+              "rubygems/ext"                            => :after,
+              "rubygems/package"                        => :after,
+              "rubygems/remote_fetcher"                 => :after,
+              "rubygems/request"                        => :after,
+              "rubygems/request_set/gem_dependency_api" => :after,
+
+              "rspec/support"                           => :after,
+
+              "unf/normalizer_cruby"                    => :before,
+  }
+
+  PATCHES.each do |file, _|
+    unless File.exist? File.join(DIR, file) + '.rb'
+      raise "file #{file} is missing in #{DIR}"
+    end
+  end
+
+  Dir["#{DIR}/**/*.rb"].each do |file|
+    relative_path = file[(DIR.size + 1)..-4]
+    unless PATCHES.key? relative_path
+      raise "file #{relative_path} is not declared in PATCHES"
+    end
+  end
+
+  def self.find_absolute_path(relative_path)
+    $LOAD_PATH.each do |base|
+      candidate = File.join(base, relative_path) + '.rb'
+      return candidate if File.exist? candidate
+    end
+    raise "absolute path for #{relative_path} was not found in #{$LOAD_PATH}"
+  end
+
+  def self.find_relative_path(absolute_path)
+    return nil unless File.exist? absolute_path
+
+    $LOAD_PATH.each do |base|
+      if absolute_path.start_with?(base)
```

Comment:
Yeah this is a limitation the lib of the tested gem has to be in `$LOAD_PATH` so the relative paths can be found to find the patches. I don't know how to make this better. At least the error could warn about possibly missing lib in `$LOAD_PATH`.